### PR TITLE
Make library flexible to support new api functions

### PR DESCRIPTION
**Pull request by dshafer**
Adds `options` hash to `pair()` and `authenitcate()` to be able to pass arbitrary options to API methods.

Gives access to the raw API response in pairing and authentication status objects, through the `.raw` attribute

Allows callers to override `base_url`, updates demo to use this feature.

### CODE REVIEW

File: test/test_toopher_api.rb [hunk 1]
```diff
@@ -43,8 +43,27 @@ def test_create_pairing_immediate_success()
       assert(pairing.enabled == true, 'pairing not enabled')
       assert(pairing.user_id == '1', 'bad user id')
       assert(pairing.user_name == 'user', 'bad user name')
+      assert(pairing.raw['user']['name'] == 'user', 'could not access raw data')
+  end
+
+  def test_create_pairing_with_optional_arg()
+    stub_http_request(:post, "https://toopher-api.appspot.com/v1/pairings/create").
```
Comment (smholloway) [inline test/test_toopher_api.rb]: Do we use `api.toopher.com` or `toopher-api.appspot.com` in tests?

File: test/test_toopher_api.rb [hunk 2]
```diff
@@ -92,6 +111,27 @@ def test_create_authentication_with_no_action()
     assert(auth.terminal_name == 'term name', 'wrong auth terminal name')
   end
 
+  def test_create_authentication_with_optional_arg()
+    stub_http_request(:post, "https://toopher-api.appspot.com/v1/authentication_requests/initiate").
```
Comment (smholloway) [inline test/test_toopher_api.rb]: `api.toopher.com` here?

Comment (dshafer) [inline test/test_toopher_api.rb]: actually should probably do something like `http://testonly.url`

**Comment by smholloway** [comment]
Looks good.

File: test/test_toopher_api.rb (hunk 2 above)
Comment (trdarr) [inline test/test_toopher_api.rb]: Use https://toopher.test/? (See: [RFC 2606](https://tools.ietf.org/html/rfc2606#section-2).)

**Comment by smholloway** [comment]
I took @trdarr 's suggestion and changed the tests to use `toopher.test`, then updated the Toopher API calls to use  `toopher.test` as their `base_url`. 

Sorry to commit on top of you :+1:

**Comment by dshafer** [comment]
I like it.  :shipit:

**Pull request merged**